File: Day 8/Day 8.cpp
#include <vector>
#include <string>
#include <fstream>
#include <sstream>
#include <iostream>

#include "stringsplit.h"

enum class opcode {
	acc,
	jmp,
	nop
};

std::string opcode_to_string(opcode op) {
	if (op == opcode::acc) return "acc";
	else if (op == opcode::jmp) return "jmp";
	else return "nop";
}

struct Instruction {
	opcode opcode;
	int arg;
};

class Console {
public:
	Console() : m_accumulator(0), m_program_counter(0) {}
	int accumulator() { return m_accumulator; }
	int program_counter() { return m_program_counter; }
	void execute_command(const Instruction& instruction) {
		switch (instruction.opcode) {
		case opcode::acc:
			acc(instruction.arg);
			break;
		case opcode::jmp:
			jmp(instruction.arg);
			break;
		case opcode::nop:
			nop();
			break;
		}
	}
	bool run_all(const std::vector<Instruction>& code) {
		reset();
		std::vector<bool> visited(code.size(), false);
		while (m_program_counter < code.size()) {
			if (visited[m_program_counter]) return false;
			visited[m_program_counter] = true;
			execute_command(code[m_program_counter]);
		}
		return true;
	}
	
	void fix_code(std::vector<Instruction> code) {  // intentional passing by value for safety reasons
		for (auto& cmd : code) {
			opcode orig = cmd.opcode;
			switch (cmd.opcode) {
			case opcode::jmp:
				cmd.opcode = opcode::nop;
				break;
			case opcode::nop:
				cmd.opcode = opcode::jmp;
				break;
			default:
				continue;
			}
			if (run_all(code) == true) {
				std::cout << "Original: " << opcode_to_string(orig) << " " << cmd.arg
					<< "\tReplacement: " << opcode_to_string(cmd.opcode) << " " << cmd.arg
					<< "    Final program counter: " << m_program_counter << "\n";
			}
			cmd.opcode = orig;
		}
	}
	void reset() {
		m_accumulator = 0;
		m_program_counter = 0;
	}
private:
	int m_accumulator;
	int m_program_counter;
	void acc(int arg) {
		m_accumulator += arg;
		++m_program_counter;
	}
	void jmp(int arg) {
		m_program_counter += arg;
	}
	void nop() {
		++m_program_counter;
	}

};

std::vector<Instruction> parse_input(const std::string& filename) {
	std::vector<Instruction> code;
	std::ifstream file(filename);
	std::string line;
	while (std::getline(file, line)) {
		auto raw_instruction = split(strip(line), " ");
		int arg;
		opcode op;
		if (raw_instruction[0] == "acc") op = opcode::acc;
		else if (raw_instruction[0] == "jmp") op = opcode::jmp;
		else op = opcode::nop;
		std::stringstream sstream(raw_instruction[1]);
		sstream >> arg;
		code.emplace_back(Instruction{ op, arg });
	}
	return code;
}

int main() {
	auto code = parse_input("input.txt");
	Console console;
	console.run_all(code);
	std::cout << "Stopped at program counter " << console.program_counter() << " with accumulator at " << console.accumulator() << "\n";
	console.fix_code(code);
	std::cout << "Accumulator was at " << console.accumulator() << " when the program terminated.\n";
}

File: Day 8/Day_8.cs
using System;
using System.Collections.Generic;

namespace Day_8
{
    enum Opcode
    {
        acc,
        jmp,
        nop
    }

    struct Instruction
    {
        public Instruction(Opcode op, int arg)
        {
            Opcode = op;
            Arg = arg;
        }
        public Opcode Opcode;
        public int Arg;
    }

    class GameConsole
    {
        public GameConsole()
        {
            Accumulator = 0;
            ProgramCounter = 0;
        }
        public int Accumulator { get; private set; }
        public int ProgramCounter { get; private set; }

        public void ExecuteCommand(in Instruction instruction)
        {
            switch (instruction.Opcode)
            {
                case Opcode.acc:
                    Acc(instruction.Arg);
                    break;
                case Opcode.jmp:
                    Jmp(instruction.Arg);
                    break;
                case Opcode.nop:
                    Nop();
                    break;
            }
        }

        public bool RunAll(Instruction[] code)
        {
            Reset();
            bool[] visited = new bool[code.Length];
            while (ProgramCounter < code.Length)
            {
                if (visited[ProgramCounter]) return false;
                visited[ProgramCounter] = true;
                ExecuteCommand(code[ProgramCounter]);
            }
            return true;
        }

        public void FixCode(Instruction[] code)
        {
            // It didn't allow assigning with foreach... C++ allows it with range-based for using non-const ref...
            for (int i = 0; i < code.Length; i++)
            {
                Opcode orig = code[i].Opcode;
                switch(code[i].Opcode)
                {
                    case Opcode.jmp:
                        code[i].Opcode = Opcode.nop;
                        break;
                    case Opcode.nop:
                        code[i].Opcode = Opcode.jmp;
                        break;
                    default:
                        continue;
                }
                if (RunAll(code) == true)
                {
                    Console.WriteLine("Original: " + orig + " " + code[i].Arg + "\tReplacement: " + code[i].Opcode + " " + code[i].Arg
                        + "\tFinal Program Counter: " + ProgramCounter);
                    break;
                }
                code[i].Opcode = orig;
            }
        }
        public void Reset()
        {
            Accumulator = 0;
            ProgramCounter = 0;
        }

        private void Acc(int arg)
        {
            Accumulator += arg;
            ++ProgramCounter;
        }

        private void Jmp(int arg)
        {
            ProgramCounter += arg;
        }

        private void Nop()
        {
            ++ProgramCounter;
        }
    }
    class Day_8
    {
        static Instruction[] ParseInput(string filename)
        {
            string raw_contents = System.IO.File.ReadAllText(filename);
            string[] lines = raw_contents.Trim().Split("\n");
            var code = new List<Instruction>();
            foreach (var line in lines)
            {
                var raw_instruction = line.Split(" ");
                int arg;
                Opcode op;
                if (raw_instruction[0] == "acc") op = Opcode.acc;
                else if (raw_instruction[0] == "jmp") op = Opcode.jmp;
                else op = Opcode.nop;
                arg = int.Parse(raw_instruction[1]);
                code.Add(new Instruction(op, arg));
            }

            return code.ToArray();
        }
        static void Main()
        {
            var code = ParseInput("input.txt");
            var console = new GameConsole();
            console.RunAll(code);
            Console.WriteLine("Stopped at program counter " + console.ProgramCounter 
                + " with accumulator at " + console.Accumulator);
            console.FixCode(code);
            Console.WriteLine("Accumulator was at " + console.Accumulator + " when the program terminated.");
        }
    }
}
